A43-sdk-authorization.md

```diff
@@ -0,0 +1,424 @@
+A43: SDK authorization
+
+* Author(s): [Ashitha Santhosh](https://github.com/ashithasantosh)
+* Approver: 
+* Status: Draft {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2021-06-17
+* Discussion at: (filled after thread exists)
+
+## Abstract
+
+gRPC is currently implementing an open-source standard authorization solution
+based on Envoy RBAC. This proposal discusses how this framework would be made 
+available to non-xDS OSS gRPC users to perform per-RPC authorization checks.
+
+## Background
+
+Authentication verifies the authority of the request. Authorization ensures the 
+authenticated authority has access to the requested resource.
+
+gRPC offers built-in authentication mechanisms like TLS, ALTS and even allow
+users to plug in their own authentication systems. However, gRPC doesn't have any
+standard authorization solution available. If a user wanted to perform per RPC
+authorization check, they have to implement their own solution in the application.
+With standard gRPC authorization solution (under implementation), users can simply
+provide a policy configuration for their service, then the server verifies
+whether the client is authorized to access the requested service, and for which
+RPC methods. This differs from client-side authorization (also known as hostname
+verification) where client verifies server's identity.
+
+We have two different use-cases for authorization-
+1. gRPC Proxyless Service Mesh (or xDS authz) - Here Istio or Traffic Director
+   control plane sends Envoy RBAC policy configs to xDS enabled gRPC servers for
+   authorization
+2. SDK authorization (or non-xDS authz) - Authorization is made available to
+   non-xDS OSS gRPC users like network switches.
+
+This document discusses details of SDK authorization use-case. xDS authorization
+will be convered in a different proposal.
+
+## Related Proposals:
+
+* [A41](https://github.com/grpc/proposal/pull/237): xds RBAC Support
+
+## Proposal
+
+### User facing Authorization Policy
+
+gRPC SDK authorization policy is the user facing policy language that allows
+service owners and/or security admins to enable per-RPC authorization checks.
+
+SDK authorization policy guarantees stability and backward compatibility. So if
+a user writes a config, it will always be supported by gRPC unless the policy
+undergoes major versioning change like adding required fields. The policy is 
+extendable, we may add new fields in the future as new requirements are introduced, 
```

Comment:
Hmm... if a new field is added and a rule uses it on an older version of gRPC, then the restriction will be ignored. Seems that could cause trouble for allow rules. Do we want to do anything to address that?

Reply:
If a new field is added and user creates a config which includes this new field, and we run it against an older version of gRPC, that field will be ignored in the evaluation. Required or optional, this is undesired behavior. But I thought this should be user's responsibility to make sure they run against the right version. So if they want to use new fields, they would have to upgrade to the required version. I couldn't think of a way to address this. Do you have any suggestions?
Also you mentioned allow rules specifically, but we would have the same concern for deny rules as well?

This paragraph mainly talks about the life of a policy. So as long as we didn't introduce any "required" fields (we won't be deleting any fields), the same policy could be used across different gRPC versions.

Reply:
Deny would also have trouble, but they would fail things that shouldn't be failed. That is annoying, but isn't a security problem.

The only idea I have to resolve this problem other than looking for unknown fields is have a "version" field that starts at 1. When we add new fields that must be understood, then any config containing those fields would require version to be 2. But if these are hand-edited it doesn't help too much.

@yihuazhang, have any thoughts?

Reply:
Thanks for bringing this up, and I think there are two separate problems here:

1. Detect the problem of a "new" SDK policy being evaluated by an "old" gRPC library.
2. Action to be taken when 1) happens - either ignore the new field and continue with policy evaluation or disallow the use of SDK. 

For 1), one option I can think of is to ship an SDK JSON schema within each gRPC release, that should be considered as a source of truth for the specific gRPC release, and use it to validate a user provided SDK policy. It does not take any user provided inputs (e.g., version number) during validation, and the approach (i.e., JSON schema validation) seems pretty well adopted. 

For 2), from the security perspective, I prefer to disallow the use of SDK. By silently ignoring the new field, it may give the service owner false impression that SDK is functioning as expected. More importantly, it may allow the requests that should be rejected. Another option is to log an error when the problem happens, but it may not help much if the service owner does not consume the log in a correct way (e.g., never check the log).

Reply:
For 1), another option is to code the translator parsing logic to represent the SDK schema it currently supports, in which case there will be no need to ship the schema with its gRPC release. This is what gRPC Go currently does. @dfawley F.Y.I.

Reply:
Currently both C++ and Go RBAC translator ignores fields other than the ones mentioned in JSON schema. It could be new/unknown fields. So I will look into the JSON library and figure out on how we can error out and mark the policy as invalid if there are fields other than what's expected in the current version. Will update the proposal with the details.

Reply:
The current idea is that translator will mark the policy as invalid, if the policy contains fields that are unknown to the current JSON schema. Updated the proposal accordingly.
So far I checked C++ and Go, we can support this behavior.

Though this means, we won't have backward compatibility. We won't be able to use a policy with older gRPC versions.